internal/ethapi/api.go

```diff
@@ -1174,19 +1174,25 @@ func (s *PublicBlockChainAPI) GetPreviousCheckpointFromEpoch(ctx context.Context
 	epoch := s.b.ChainConfig().XDPoS.Epoch
 
 	if epochNum == rpc.LatestEpochNumber {
-		blockNumer := s.b.CurrentBlock().Number().Uint64()
-		diff := blockNumer % epoch
-		// checkpoint number
-		checkpointNumber = blockNumer - diff
-		epochNum = rpc.EpochNumber(checkpointNumber / epoch)
-		if diff > 0 {
-			epochNum += 1
+		blockNumer := s.b.CurrentBlock().Number()
+		if engine, ok := s.b.GetEngine().(*XDPoS.XDPoS); ok {
+			var err error
+			var currentEpoch uint64
+			checkpointNumber, currentEpoch, err = engine.GetCurrentEpochSwitchBlock(s.chainReader, blockNumer)
+			if err != nil {
+				log.Error("[GetPreviousCheckpointFromEpoch] Fail to get GetCurrentEpochSwitchBlock for current checkpoint block", "block", blockNumer, "err", err)
+				return 0, epochNum
+			}
+
+			epochNum = rpc.EpochNumber(currentEpoch)
```

Comment:
The function name is `GetPreviousCheckpointFromEpoch`, but `epochNum` is the return value of `engine.GetCurrentEpochSwitchBlock`.